### ISSUE TITLE
fix(compute): replace finite CUDA reduction sentinels with true bounds

### ISSUE THREAD

**Issue by anthony-chaudhary** [anthony-chaudhary/fak#12173]
# fix(compute): replace finite CUDA reduction sentinels with true bounds

```routing
lane: internal/compute
paths: ["internal/compute/cuda_kernels.cu", "internal/compute/cuda_flash_test.go", "internal/compute/cuda_argmax_test.go"]
expected_steps: 6
```

## Current state

Four CUDA reductions initialize a maximum with `-1e30f`: the retained naive
attention baseline, decode flash attention, DSA sparse attention, and argmax.
IEEE-754 float32 values below `-1e30` are finite and valid. Attention then keeps
an empty normalizer and can return zeros; argmax can retain index zero even when
a later element is larger.

A physical sm89 witness reached this state through the decode flash ABI using
finite Q/K values whose dot products were below the sentinel: expected norm
`1.92133251`, CUDA norm `0`, cosine `0`, and maximum absolute error `0.125`.
The neighboring split-KV path was fixed under #11100, leaving these independent
production and baseline reductions exposed.

## Problem Frame

- Centrality: Core compute correctness; these reductions choose attention probabilities and output tokens.
- P1 Context: Advanced; ordinary random tests do not cross the assumed score floor, so the defect hides behind otherwise stable softmax algebra.
- P2 Net value: advanced - one shared invariant removes silent zero vectors and incorrect token selection without changing supported APIs.
- P3 Adaptation: preserved - change only empty/max initializers and add focused extreme-finite regressions.
- P4 Operations: Preserved; no allocation, launch geometry, scheduling, or transfer behavior changes.

## Core through-line

Replace every CUDA maximum accumulator that semantically means “no finite value
seen yet” with true negative infinity, then prove attention and argmax accept
finite inputs below `-1e30` on a physical CUDA device.

## Why this is next

The #11100 physical regression proved the same invalid assumption in split-KV
attention. Its independent sequential flash baseline also returned a zero vector,
so the remaining production exposure is witnessed rather than hypothetical.

## Parent context

Follow-up discovered while physically verifying #11100.

## Working spine

Extreme finite Q/K or logits -> CUDA reduction initializer -> online softmax or
argmax update -> correct nonzero attention output or maximum index -> device receipt.

## Gold-plating boundary

Do not redesign attention tiling, change reduction order, tune occupancy, alter
NaN policy, or sweep Vulkan/Metal shaders in this leaf. Do not conflate masked
`-1e300` double values with empty float32 reduction state.

## Done condition

- [ ] Naive attention, decode flash attention, DSA sparse attention, and CUDA argmax use true negative infinity for empty/max state.
- [ ] A physical CUDA attention witness with finite scores below `-1e30` matches an independent CPU oracle at cosine at least 0.99999 and maximum absolute delta at most `1e-5`.
- [ ] A CUDA argmax witness over all-finite values below `-1e30` returns the actual maximum index.
- [ ] Source contracts prevent reintroduction of finite reduction sentinels.
- [ ] Focused tests, full compute tests, vet, sm80/sm89/sm90 compilation, boundary, and leak audits pass.

## Definition of Done

- [ ] All four CUDA maximum initializers represent a true empty bound.
- [ ] Attention and argmax edge cases execute correctly on a physical CUDA device.
- [ ] Default-CI source contracts and the full compute package remain green.

## Acceptance gate

The physical attention and argmax edge witnesses pass, all three supported CUDA
architectures compile, and the focused/full package checks exit zero.

## Closure binding

The DCO-signed resolving commit cites the created public issue and carries
`(fak compute)`.

## Witness

```bash
go test ./internal/compute -run 'CUDA.*(Attention|Argmax).*Extreme|FiniteReductionSentinel' -count=1
go test ./internal/compute -count=1
go vet ./internal/compute/...
```

Physical CUDA execution must additionally invoke the real exported attention
and argmax ABIs with the edge inputs above; source inspection alone is not a
device witness.

## Likely files

- `internal/compute/cuda_kernels.cu:2057` (`k_attention` block maximum)
- `internal/compute/cuda_kernels.cu:2142` (`k_flash_attention` online softmax)
- `internal/compute/cuda_kernels.cu:2571` (`k_dsa_sparse_attend` online softmax)
- `internal/compute/cuda_kernels.cu:2740` (`k_argmax` maximum)
- `internal/compute/cuda_flash_test.go`
- `internal/compute/cuda_argmax_test.go`

## Blast radius and affected lanes

- Affected: CUDA attention and argmax inside `internal/compute`.
- Unaffected: CPU, Vulkan, Metal, model loading, serving APIs, and kernel launch geometry.
- Fallback: existing normal-range paths remain unchanged; unsupported CUDA builds continue to exclude tagged device tests.

## Lane

internal/compute

## Work estimate

Estimate: 2 points.

## Overall completion contribution

Contribution: 2/2 points.

## Completion standard

development

## Target operating envelope

physical CUDA defect-reproduction coverage: = 100 percent

## Witnessed operating envelope

physical CUDA defect-reproduction coverage: = 100 percent

**Comment by anthony-chaudhary** [comment]
Landed on public `main` in `88f43143b39a313befced33d17a105a20e6bc2b7`.

## Before/after device witness

The discovery witness on an RTX 4070 Laptop GPU (sm89) fed finite Q/K values whose attention scores were below `-1e30`:

- before: expected output norm `1.92133251`, CUDA flash output norm `0`, cosine `0`, maximum absolute error `0.125`
- after: flash cosine `1`, maximum absolute error `0`

The post-fix direct CUDA harness exercised all four corrected reductions from the current translation unit:

- fused decode flash attention: cosine `1`, maximum absolute error `0`
- retained naive attention: cosine `1`, maximum absolute error `0`
- DSA sparse attention: cosine `1`, maximum absolute error `0`
- argmax over `[-3e30, -2e30, -2.5e30, -4e30]`: returned index `1`, as required

The attention oracle is the closed-form uniform mean for equal finite scores, independent of the CUDA recurrence. The change replaces only empty/max initializers with true negative infinity; launch geometry, allocation, and reduction order are unchanged.

## Verification

- `go test ./internal/compute -count=1 -timeout=120s` — PASS (`117.164s`)
- focused default-CI source contract — PASS
- `go vet ./internal/compute/...` — PASS
- whole CUDA translation unit compiled for sm80, sm89, and sm90
- physical sm89 execution — PASS for flash, naive attention, DSA, and argmax
- issue discoverability audit — PASS, dispatchable
- staged public leak audit — PASS
- private/public five-gate boundary audit — PASS
- `fak validate --mine` emitted no output for 90 seconds and was interrupted; its direct package, device, architecture, leak, and boundary components above completed independently
